### PR TITLE
Add Registration ID to Enriched TD

### PR DESCRIPTION
**Pull request by farshidtz**
Related to #149.

I added the URL inside the registration object for now to be consistent with the other registration attributes. We could move them all to the top level and pick more suitable names.

```json
{
    "@context": [
        "http://www.w3.org/ns/td",
        "https://w3c.github.io/wot-discovery/context/discovery-context.jsonld"
    ],
    "id": "urn:example:simple-td",
    "security": "basic_sc",
    "securityDefinitions": {
        "basic_sc": {
            "scheme": "basic"
        }
    },
    "tdd:registration": {
        "tdd:created": "2021-01-19T17:02:00Z",
        "tdd:href": "https://tdd.example.com/td/urn:example:simple-td",
        "tdd:modified": "2021-01-19T17:02:00Z"
    },
    "title": "Simple TD"
}
```

Also, I made it explicit that retrieved TDs will be enriched. This is necessary to identify Anonymous TDs in listing and search APIs.

***
<a href="https://pr-preview.s3.amazonaws.com/farshidtz/wot-discovery/pull/152.html" title="Last updated on Apr 19, 2021, 11:08 AM UTC (6030b4c)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/wot-discovery/152/fab5087...farshidtz:6030b4c.html" title="Last updated on Apr 19, 2021, 11:08 AM UTC (6030b4c)">Diff</a>

### CODE REVIEW

**Comment by farshidtz** [comment]
Looking at the sorting and notification requirements, I think adding an `id` instead of `href` will be more appropriate. This will be same as what we originally agreed on but removed because `@id` of `tdd:registration` object was semantically the id of registration object and nothing related to the TD. One solution to alias `tdd:id` to something like thingRegistrationID instead of `@id`. 

Reasons for `id` instead of `href`:
- Sorting by a shorter string will be computationally less complex than sorting by `href`.
- Construction of an absolute URL requires knowledge of the fully qualified domain name at the time of deployment and will prevent serving via both local and public endpoints.

```json
{
    "@context": [
        "http://www.w3.org/ns/td",
        "https://w3c.github.io/wot-discovery/context/discovery-context.jsonld"
    ],
    "id": "urn:example:simple-td", // anonymous TDs will not have this
    "security": "basic_sc",
    "securityDefinitions": {
        "basic_sc": {
            "scheme": "basic"
        }
    },
    "tdd:registration": {
        "tdd:created": "2021-01-19T17:02:00Z",
        "tdd:href": "https://tdd.example.com/td/urn:example:simple-td",
        "tdd:id": "urn:example:simple-td", // id instead of href?
        "tdd:modified": "2021-01-19T17:02:00Z"
    },
    "title": "Simple TD"
}
```

**Comment by farshidtz** [comment]
I just had a call with @AndreaCimminoArriaga to discuss the proposal. We came up with:

Context file based on what we already have in the ontology:
```json
{
  "@context": {
    "discovery": "https://www.w3.org/2021/wot/discovery#",
    "DirectoryDescription": {
      "@id": "discovery:DirectoryDescription"
    },
    "LinkDescription": {
      "@id": "discovery:LinkDescription"
    },
    "RegistrationInformation": {
      "@id": "discovery:RegistrationInformation"
    },
    "id": "@id",
    "type": "@type",
    "registration": "discovery:hasRegistrationInformation",
    "created": "dc:created",
    "modified": "dc:modified"
  }
}
```

Known TD:
```json
{
    "@context": [
        "http://www.w3.org/ns/td",
        "https://w3c.github.io/wot-discovery/context/discovery-context.jsonld"
    ],
    "id": "urn:example:known-td",
    "security": "basic_sc",
    "securityDefinitions": {
        "basic_sc": {
            "scheme": "basic"
        }
    },
    "registration": {
        "created": "2021-01-19T17:02:00Z",
        "id": "https://tdd.example.com/td/urn:example:known-td", 
        "modified": "2021-01-19T17:02:00Z"
    },
    "title": "Known TD"
}
```

Anonymous TD:
```json
{
    "@context": [
        "http://www.w3.org/ns/td",
        "https://w3c.github.io/wot-discovery/context/discovery-context.jsonld"
    ],
    "security": "basic_sc",
    "securityDefinitions": {
        "basic_sc": {
            "scheme": "basic"
        }
    },
    "registration": {
        "created": "2021-01-19T17:02:00Z",
        "id": "https://tdd.example.com/td/urn:uuid:ab90172b-c3b7-4f71-9ea0-2dc7cf9b796f", 
        "modified": "2021-01-19T17:02:00Z"
    },
    "title": "Anonymous TD"
}
```

We can move all the registration attributes to TD itself and give them different names. That way, we would become inline with @vcharpenay's suggestion at https://github.com/w3c/wot-discovery/issues/149#issuecomment-817746786. Signature verification can still work either by specifying what is signed (https://github.com/w3c/wot-thing-description/issues/940#issuecomment-816570048) or by removing (unenriching) the TD before validation. Moreover, this makes the notification object simpler (more on that in a future PR).

**Comment by farshidtz** [comment]
There is a recommended way to add local identifiers to blank nodes (in this case the Anonymous TD):
https://json-ld.org/spec/latest/json-ld/#identifying-blank-nodes

An identifier that is not an absolute URL also brings the following benefits:
- Sorting by a shorter string will be computationally less complex than sorting by url. We can still recommend servers to sort by the variable part of it, but we need to describe that variable and maybe have it in the ontology too.
- Construction of an absolute URL requires knowledge of the fully qualified domain name (FQDN) at the time of deployment and will prevent serving via both local or more than one public (proxied) endpoints.

File: index.html [hunk 1]
```diff
@@ -773,8 +824,10 @@ <h4>Creation</h4>
                                     and a Location header containing a system-generated identifier for the TD.
                                 </span>
                                 <span class="rfc2119-assertion" id="tdd-reg-create-anonymous-td-generated-id"></span>
-                                    The identifier SHOULD be a Version 4 UUID URN [[RFC4122]].
+                                    The identifier SHOULD be a UUID Version 4 [[RFC4122]].
                                 </span>
+                                That is a random or pseudo-random number which does not carry unintended information
+                                about the host or the resource.
```
Comment (farshidtz) [inline index.html:830]: @mmccool should we add some recommendation about generation of secure identifiers in the privacy and security section? In particular, UUID versions other than v4 may need to be avoided.

Comment (mmccool) [reply index.html:830]: please create an issue about this in wot-security and we will discuss

**Pull request merged**